In [1]:
import pandas as pd
data=pd.read_csv("C:/Users/User/Documents/IMDB Dataset.csv")
data

review sentiment
0      One of the other reviewers has mentioned that ...  positive
1      A wonderful little production. <br /><br />The...  positive
2      I thought this was a wonderful way to spend ti...  positive
3      Basically there's a family where a little boy ...  negative
4      Petter Mattei's "Love in the Time of Money" is...  positive
...                                                  ...       ...
49995  I thought this movie did a down right good job...  positive
49996  Bad plot, bad dialogue, bad acting, idiotic di...  negative
49997  I am a Catholic taught in parochial elementary...  negative
49998  I'm going to have to disagree with the previou...  negative
49999  No one expects the Star Trek movies to be high...  negative

[50000 rows x 2 columns]

    there's two columns review and sentiment, for example we get one row value

In [2]:
review_1=data.loc[0,'review']
review_1

"One of the other reviewers has mentioned that after watching just 1 Oz episode you'll be hooked. They are right, as this is exactly what happened with me.<br /><br />The first thing that struck me about Oz was its brutality and unflinching scenes of violence, which set in right from the word GO. Trust me, this is not a show for the faint hearted or timid. This show pulls no punches with regards to drugs, sex or violence. Its is hardcore, in the classic use of the word.<br /><br />It is called OZ as that is the nickname given to the Oswald Maximum Security State Penitentary. It focuses mainly on Emerald City, an experimental section of the prison where all the cells have glass fronts and face inwards, so privacy is not high on the agenda. Em City is home to many..Aryans, Muslims, gangstas, Latinos, Christians, Italians, Irish and more....so scuffles, death stares, dodgy dealings and shady agreements are never far away.<br /><br />I would say the main appeal of the show is due to the fa

    the text is not clean, there're are syntax html and noise in word like punctuatuons and many more. we want to remove them

In [3]:
#remove html tags
from bs4 import BeautifulSoup
clean_text=BeautifulSoup(review_1,"html.parser").get_text()
clean_text

"One of the other reviewers has mentioned that after watching just 1 Oz episode you'll be hooked. They are right, as this is exactly what happened with me.The first thing that struck me about Oz was its brutality and unflinching scenes of violence, which set in right from the word GO. Trust me, this is not a show for the faint hearted or timid. This show pulls no punches with regards to drugs, sex or violence. Its is hardcore, in the classic use of the word.It is called OZ as that is the nickname given to the Oswald Maximum Security State Penitentary. It focuses mainly on Emerald City, an experimental section of the prison where all the cells have glass fronts and face inwards, so privacy is not high on the agenda. Em City is home to many..Aryans, Muslims, gangstas, Latinos, Christians, Italians, Irish and more....so scuffles, death stares, dodgy dealings and shady agreements are never far away.I would say the main appeal of the show is due to the fact that it goes where other shows wo

    There's no html text in the sentence. but there'es elipsis (three dots in the text), we should remove unless there're important test for meaning

In [4]:
#\.\.\.|\.\. remove .... and .. in text and change " " spacing
import re
clean_text_elip=re.sub(r"\.\.\.|\.\."," ",clean_text)
clean_text_elip

"One of the other reviewers has mentioned that after watching just 1 Oz episode you'll be hooked. They are right, as this is exactly what happened with me.The first thing that struck me about Oz was its brutality and unflinching scenes of violence, which set in right from the word GO. Trust me, this is not a show for the faint hearted or timid. This show pulls no punches with regards to drugs, sex or violence. Its is hardcore, in the classic use of the word.It is called OZ as that is the nickname given to the Oswald Maximum Security State Penitentary. It focuses mainly on Emerald City, an experimental section of the prison where all the cells have glass fronts and face inwards, so privacy is not high on the agenda. Em City is home to many Aryans, Muslims, gangstas, Latinos, Christians, Italians, Irish and more .so scuffles, death stares, dodgy dealings and shady agreements are never far away.I would say the main appeal of the show is due to the fact that it goes where other shows would

**missing space after .**: find the word using ?<= (before) and ?=(after) in regex. write in () like grouping in write in [] matching that you want search

In [5]:
find_word=re.findall(r"(\b[a-zA-Z\s]+)\.(?=[a-zA-Z])", clean_text_elip)
#  Irish and more .so scuffles, 
#f the word.It is called OZ
find_word

['as this is exactly what happened with me',
 'in the classic use of the word',
 'Irish and more ',
 'dodgy dealings and shady agreements are never far away',
 'you may become comfortable with what is uncomfortable viewing ']

r"(\b[a-zA-Z\s]+)\.(?=[a-zA-Z]) explanation:

- we want find sentence so using \b (boundary text) 
-   "+" means that one and more letter, so we can get full words, without it we get single letter, + heplp us extract full words 
- \s is space


In [6]:
#add space after . if after . in the text is not space
clean_text2=re.sub(r"(?<=[a-zA-Z\s])\.(?=[a-zA-Z])",". ",clean_text_elip)
clean_text2

"One of the other reviewers has mentioned that after watching just 1 Oz episode you'll be hooked. They are right, as this is exactly what happened with me. The first thing that struck me about Oz was its brutality and unflinching scenes of violence, which set in right from the word GO. Trust me, this is not a show for the faint hearted or timid. This show pulls no punches with regards to drugs, sex or violence. Its is hardcore, in the classic use of the word. It is called OZ as that is the nickname given to the Oswald Maximum Security State Penitentary. It focuses mainly on Emerald City, an experimental section of the prison where all the cells have glass fronts and face inwards, so privacy is not high on the agenda. Em City is home to many Aryans, Muslims, gangstas, Latinos, Christians, Italians, Irish and more . so scuffles, death stares, dodgy dealings and shady agreements are never far away. I would say the main appeal of the show is due to the fact that it goes where other shows w

In [7]:
#remove space before .
clean_text3=re.sub(r"\s\.",".",clean_text2)
clean_text3

"One of the other reviewers has mentioned that after watching just 1 Oz episode you'll be hooked. They are right, as this is exactly what happened with me. The first thing that struck me about Oz was its brutality and unflinching scenes of violence, which set in right from the word GO. Trust me, this is not a show for the faint hearted or timid. This show pulls no punches with regards to drugs, sex or violence. Its is hardcore, in the classic use of the word. It is called OZ as that is the nickname given to the Oswald Maximum Security State Penitentary. It focuses mainly on Emerald City, an experimental section of the prison where all the cells have glass fronts and face inwards, so privacy is not high on the agenda. Em City is home to many Aryans, Muslims, gangstas, Latinos, Christians, Italians, Irish and more. so scuffles, death stares, dodgy dealings and shady agreements are never far away. I would say the main appeal of the show is due to the fact that it goes where other shows wo

In [8]:
#change uppercase when after .  is lowercase
#group (1) only first word. group(1) get info from [a-z] or get first letter after .
def upper(x):
    return ". "+x.group(1).upper()
clean_text4=re.sub(r"\.\s([a-z])",upper,clean_text3)
clean_text4

"One of the other reviewers has mentioned that after watching just 1 Oz episode you'll be hooked. They are right, as this is exactly what happened with me. The first thing that struck me about Oz was its brutality and unflinching scenes of violence, which set in right from the word GO. Trust me, this is not a show for the faint hearted or timid. This show pulls no punches with regards to drugs, sex or violence. Its is hardcore, in the classic use of the word. It is called OZ as that is the nickname given to the Oswald Maximum Security State Penitentary. It focuses mainly on Emerald City, an experimental section of the prison where all the cells have glass fronts and face inwards, so privacy is not high on the agenda. Em City is home to many Aryans, Muslims, gangstas, Latinos, Christians, Italians, Irish and more. So scuffles, death stares, dodgy dealings and shady agreements are never far away. I would say the main appeal of the show is due to the fact that it goes where other shows wo

In [9]:
find_word2=re.findall(r"(?<=\s)(?<!\.\s)\b[A-Z][a-z0-9]+", clean_text4)
find_word2

['Oz',
 'Oz',
 'Oswald',
 'Maximum',
 'Security',
 'State',
 'Penitentary',
 'Emerald',
 'City',
 'City',
 'Aryans',
 'Muslims',
 'Latinos',
 'Christians',
 'Italians',
 'Irish',
 'Oz',
 'Watching',
 'Oz']

In [10]:
import spacy
nlp=spacy.load("en_core_web_sm")
doc=nlp(" ".join(find_word2))
for i in doc:
    print(i.text+":"+i.pos_)

Oz:PROPN
Oz:PROPN
Oswald:PROPN
Maximum:PROPN
Security:PROPN
State:PROPN
Penitentary:PROPN
Emerald:PROPN
City:PROPN
City:PROPN
Aryans:PROPN
Muslims:PROPN
Latinos:PROPN
Christians:PROPN
Italians:PROPN
Irish:PROPN
Oz:PROPN
Watching:PROPN
Oz:PROPN


    since its dificult to distingusih word noun or not we using spacy

In [11]:
### import spacy
nlp=spacy.load("en_core_web_sm")
#make doc
doc=nlp(clean_text4)
#token sentences
sentence=[i.text for i in doc.sents]
sentence

["One of the other reviewers has mentioned that after watching just 1 Oz episode you'll be hooked.",
 'They are right, as this is exactly what happened with me.',
 'The first thing that struck me about Oz was its brutality and unflinching scenes of violence, which set in right from the word GO.',
 'Trust me, this is not a show for the faint hearted or timid.',
 'This show pulls no punches with regards to drugs, sex or violence.',
 'Its is hardcore, in the classic use of the word.',
 'It is called OZ as that is the nickname given to the Oswald Maximum Security State Penitentary.',
 'It focuses mainly on Emerald City, an experimental section of the prison where all the cells have glass fronts and face inwards, so privacy is not high on the agenda.',
 'Em City is home to many Aryans, Muslims, gangstas, Latinos, Christians, Italians, Irish and more.',
 'So scuffles, death stares, dodgy dealings and shady agreements are never far away.',
 "I would say the main appeal of the show is due to t

# SCORING SENTIMENT

- lexicon based (+positive, -negative, and netral)
- Machine learning approach

**Machine Learning Approach** : using BERT, its deep learning like brain human who can recognize context of the text


**1. Embbeding** : convert raw text into number

In [12]:
from transformers import BertTokenizer, BertModel
#define tokenizer
tokenizer=BertTokenizer.from_pretrained('bert-base-uncased')
#token the sentence
sentence_1=sentence[0]
token=tokenizer(sentence_1,return_tensors="pt")
print(token)

C:\Users\User\Anaconda3\envs\nlp_env\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


{'input_ids': tensor([[  101,  2028,  1997,  1996,  2060, 15814,  2038,  3855,  2008,  2044,
          3666,  2074,  1015, 11472,  2792,  2017,  1005,  2222,  2022, 13322,
          1012,   102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}


- retur_tensors = pt, the output is tensor array
- input_ids = 101 is penanda awal kalimat, dan 102 merupakan penanda akhir pada kalimat, kemudian setiap kata pada kalimat memiliki id
- token type id= karena inpunt nya cuma singgle input maka type id 0 semua, (token type id digunakan jka ada pairing antra context dan pertanyaan, kegunaaannya untuk membedakan tersebut
- attenti mask 1 means they are real token and should use to input the model, attention mask is 0 they are padding

**Bagaimana dengan satu paragraph**

In [15]:
from transformers import BertTokenizer, BertModel
tokenizer=BertTokenizer.from_pretrained('bert-base-uncased')
review=clean_text4
token_review=tokenizer(
    review,
    truncation=True, #if token excedd number of token
    max_length=512, #maximum tokeno
    padding="max_length",
    return_tensors="pt"


)

print(token_review["input_ids"])

tensor([[  101,  2028,  1997,  1996,  2060, 15814,  2038,  3855,  2008,  2044,
          3666,  2074,  1015, 11472,  2792,  2017,  1005,  2222,  2022, 13322,
          1012,  2027,  2024,  2157,  1010,  2004,  2023,  2003,  3599,  2054,
          3047,  2007,  2033,  1012,  1996,  2034,  2518,  2008,  4930,  2033,
          2055, 11472,  2001,  2049, 24083,  1998,  4895, 10258,  2378,  8450,
          5019,  1997,  4808,  1010,  2029,  2275,  1999,  2157,  2013,  1996,
          2773,  2175,  1012,  3404,  2033,  1010,  2023,  2003,  2025,  1037,
          2265,  2005,  1996,  8143, 18627,  2030,  5199,  3593,  1012,  2023,
          2265,  8005,  2053, 17957,  2007, 12362,  2000,  5850,  1010,  3348,
          2030,  4808,  1012,  2049,  2003, 13076,  1010,  1999,  1996,  4438,
          2224,  1997,  1996,  2773,  1012,  2009,  2003,  2170, 11472,  2004,
          2008,  2003,  1996,  8367,  2445,  2000,  1996, 17411,  4555,  3036,
          2110,  7279,  4221, 12380,  2854,  1012,  

- for this is result they're padding in a tokenize, padding is not impact of meaning but impact on model (since model processed unimportant value)

**solution 1 to handle padding** : sliding window, its like you just have 512 word to read, instead of forgetting everything you remember 50 word before tuning.

In [19]:
token_review_ch=tokenizer(
    review,
    truncation=True, #if token excedd number of token
    max_length=512, #maximum tokeno
    stride=50, #keep 50 token from previous chunk,
    return_overflowing_tokens=True,
    return_tensors="pt"


)
print(token_review_ch['input_ids'])

tensor([[  101,  2028,  1997,  1996,  2060, 15814,  2038,  3855,  2008,  2044,
          3666,  2074,  1015, 11472,  2792,  2017,  1005,  2222,  2022, 13322,
          1012,  2027,  2024,  2157,  1010,  2004,  2023,  2003,  3599,  2054,
          3047,  2007,  2033,  1012,  1996,  2034,  2518,  2008,  4930,  2033,
          2055, 11472,  2001,  2049, 24083,  1998,  4895, 10258,  2378,  8450,
          5019,  1997,  4808,  1010,  2029,  2275,  1999,  2157,  2013,  1996,
          2773,  2175,  1012,  3404,  2033,  1010,  2023,  2003,  2025,  1037,
          2265,  2005,  1996,  8143, 18627,  2030,  5199,  3593,  1012,  2023,
          2265,  8005,  2053, 17957,  2007, 12362,  2000,  5850,  1010,  3348,
          2030,  4808,  1012,  2049,  2003, 13076,  1010,  1999,  1996,  4438,
          2224,  1997,  1996,  2773,  1012,  2009,  2003,  2170, 11472,  2004,
          2008,  2003,  1996,  8367,  2445,  2000,  1996, 17411,  4555,  3036,
          2110,  7279,  4221, 12380,  2854,  1012,  